scripts/version_check.py

```diff
@@ -0,0 +1,157 @@
+import subprocess
+import sys
+import re
+from argparse import (
+    ArgumentParser,
+    Namespace,
+)
+from collections import namedtuple
+from inspect import cleandoc
+from pathlib import Path
+from shutil import which
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    Union,
+)
+
+Version = namedtuple("Version", ["major", "minor", "patch"])
+
+_SUCCESS = 0
+_FAILURE = 1
+
+# fmt: off
+_VERSION_MODULE_TEMPLATE = cleandoc('''
+# ATTENTION:
+# This file is generated, do not edit it manually!
+# If you need to change the version, do so in the project.toml, e.g. by using `poetry version X.Y.Z`.
+
+MAJOR = {major}
+MINOR = {minor}
+PATCH = {patch}
+
+VERSION = f"{{MAJOR}}.{{MINOR}}.{{PATCH}}"
+''') + "\n"
+
+
+# fmt: on
+
+
+def version_from_string(s: str) -> Version:
+    """Converts a version string of the following format major.minor.patch to a version object"""
+    major, minor, patch = (int(number, base=0) for number in s.split("."))
+    return Version(major, minor, patch)
+
+
+class CommitHookError(Exception):
+    """Indicates that this commit hook encountered an error"""
+
+
+def version_from_python_module(path: Path) -> Version:
+    """Retrieve version information from the `version` module"""
+    with open(path) as file:
+        _locals: Dict[str, Any] = {}
+        _globals: Dict[str, Any] = {}
+        exec(file.read(), _locals, _globals)
+
+        try:
+            version = _globals["VERSION"]
+        except KeyError as ex:
+            raise CommitHookError("Couldn't find version within module") from ex
+
+        return version_from_string(version)
+
+
+def version_from_poetry() -> Version:
+    poetry = which("poetry")
+    if not poetry:
+        raise CommitHookError("Couldn't find poetry executable")
+
+    result = subprocess.run([poetry, "version"], capture_output=True)
+    version = result.stdout.decode().split()[1]
+    return version_from_string(version)
+
+
+def version_from_changelog(path: Path) -> Version:
+    with open(path) as changelog:
+        content = changelog.read()
+        # Search for the FIRST pattern like: "* [0.5.0](changes_0.5.0.md)" in the changelog file.
+        # Note that we encapsulate the [(0.5.0)] with parenthesis, which tells re to return the matching string as group
+        version_match = re.search(r"\* \[([0-9]+.[0-9]+.[0-9]+)]\(\S+\)", content)
```

Comment:
Should we also check the version in the changes file?

Reply:
Maybe later. I would first like to or discuss/decide on how we generally include the change log into the sphinx based documentation, e.g.:

* using `.rst` format for change log files
* include md files in sphinx
* like in the sqla project (with a single `changelog.rst` because sphinx will provide structure anyway)